src/Kiota.Builder/Caching/DocumentCachingProvider.cs

```diff
@@ -37,30 +39,35 @@ private async Task<Stream> GetDocumentInternalAsync(Uri documentUri, string inte
     {
         var hashedUrl = BitConverter.ToString((HashAlgorithm.Value ?? throw new InvalidOperationException("unable to get hash algorithm")).ComputeHash(Encoding.UTF8.GetBytes(documentUri.ToString()))).Replace("-", string.Empty);
         var target = Path.Combine(Path.GetTempPath(), "kiota", "cache", intermediateFolderName, hashedUrl, fileName);
-        if (!File.Exists(target) || couldNotDelete)
-            return await DownloadDocumentFromSourceAsync(documentUri, target, accept, token);
+        var currentLock = _locks.GetOrAdd(target, _ => new AsyncLock());
+        using (await currentLock.LockAsync(token))
+        {// if multiple clients are being updated for the same description, we'll have concurrent download of the file without the lock
+            if (!File.Exists(target) || couldNotDelete)
+                return await DownloadDocumentFromSourceAsync(documentUri, target, accept, token);
 
-        var lastModificationDate = File.GetLastWriteTime(target);
-        if (lastModificationDate.Add(Duration) > DateTime.Now && !ClearCache)
-        {
-            Logger.LogDebug("cache file {cacheFile} is up to date and clearCache is {clearCache}, using it", target, ClearCache);
-            return File.OpenRead(target);
-        }
-        else
-        {
-            Logger.LogDebug("cache file {cacheFile} is out of date, downloading from {url}", target, documentUri);
-            try
+            var lastModificationDate = File.GetLastWriteTime(target);
+            if (lastModificationDate.Add(Duration) > DateTime.Now && !ClearCache)
             {
-                File.Delete(target);
+                Logger.LogDebug("cache file {cacheFile} is up to date and clearCache is {clearCache}, using it", target, ClearCache);
```

Comment:
global nit (optional), made it a bit more explicit. I'd like to see it even more explicit about what is cached here. Is the cached file an OpenAPI description file? If so, we should make that known in the log.

```suggestion
                Logger.LogDebug("OpenAPI description file {cacheFile} is up to date and clearCache is {clearCache}, using {cacheFile}", target, ClearCache);
```

Reply:
this is not just caching OpenAPI descriptions, it's also caching the Apis.guru index and the block list.